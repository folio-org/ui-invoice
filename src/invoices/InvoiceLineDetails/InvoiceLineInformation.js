import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import {
  Checkbox,
  Col,
  KeyValue,
  NoValue,
  Row,
  TextLink,
} from '@folio/stripes/components';
import {
  ClipCopy,
  ViewMetaData,
} from '@folio/stripes/smart-components';
import {
  AmountWithCurrencyField,
  FolioFormattedDate,
  VendorReferenceNumbersDetails,
} from '@folio/stripes-acq-components';

import { CalculatedExchangeAmount } from '../../common/components';

const InvoiceLineInformation = ({
  currency,
  exchangeRate,
  invoiceLine,
  poLine,
}) => {
  const metadata = invoiceLine?.metadata;

  return (
    <>
      {metadata && <ViewMetaData metadata={metadata} />}
      <Row>
        <Col xs={12}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.description" />}
            value={invoiceLine?.description}
          />
        </Col>
        <Col xs={3}>
          <KeyValue label={<FormattedMessage id="ui-invoice.invoiceLine.poLineNumber" />}>
            {poLine?.id
              ? (
                <>
                  <TextLink
                    rel="noopener noreferrer"
                    target="_blank"
                    to={`/orders/lines/view/${poLine.id}`}
                  >
                    {poLine.poLineNumber}
                  </TextLink>
                  <ClipCopy text={poLine.poLineNumber} />
                </>
              )
              : <NoValue />
            }
          </KeyValue>
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.invoiceLineNumber" />}
            value={invoiceLine?.invoiceLineNumber}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.details.information.status" />}
            value={invoiceLine?.invoiceLineStatus}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.subscriptionInfo" />}
            value={invoiceLine?.subscriptionInfo}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.subscriptionStart" />}
            value={<FolioFormattedDate value={invoiceLine?.subscriptionStart} />}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.subscriptionEnd" />}
            value={<FolioFormattedDate value={invoiceLine?.subscriptionEnd} />}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.comment" />}
            value={invoiceLine?.comment}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.accountNumber" />}
            value={invoiceLine?.accountNumber}
          />
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoice.accountingCode" />}
            value={invoiceLine?.accountingCode}
          />
        </Col>
        <Col xs={3}>
          <KeyValue label={<FormattedMessage id="ui-invoice.invoiceLine.subTotal" />}>
            <AmountWithCurrencyField
              amount={invoiceLine?.subTotal}
              currency={currency}
            />
          </KeyValue>
        </Col>
        <Col xs={3}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.quantity" />}
            value={invoiceLine?.quantity}
          />
        </Col>
        <Col xs={3}>
          <CalculatedExchangeAmount
            currency={currency}
            exchangeRate={exchangeRate}
            label={<FormattedMessage id="ui-invoice.invoice.details.lines.list.total.exchanged" />}
            total={invoiceLine?.total}
          />
        </Col>
        <Col xs={3}>
          <Checkbox
            checked={Boolean(invoiceLine?.releaseEncumbrance)}
            disabled
            label={<FormattedMessage id="ui-invoice.invoiceLine.releaseEncumbrance" />}
            type="checkbox"
            vertical
          />
        </Col>
        <Col xs={12}>
          <KeyValue
            label={<FormattedMessage id="ui-invoice.invoiceLine.referenceNumbers" />}
          >
            <VendorReferenceNumbersDetails referenceNumbers={invoiceLine?.referenceNumbers} />
          </KeyValue>
        </Col>
      </Row>
    </>
  );
};

InvoiceLineInformation.propTypes = {
  currency: PropTypes.string,
  exchangeRate: PropTypes.number,
  invoiceLine: PropTypes.object,
  poLine: PropTypes.object,
};

export default InvoiceLineInformation;
