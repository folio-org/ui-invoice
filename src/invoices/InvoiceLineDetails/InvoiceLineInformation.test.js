import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import '../../../test/jest/__mock__';

import { orderLine, invoiceLine } from '../../../test/jest/fixtures';

import InvoiceLineInformation from './InvoiceLineInformation';

const defaultProps = {
  invoiceLine,
  currency: 'USD',
  poLine: orderLine,
};
const renderInvoiceLineInformation = (props = defaultProps) => (render(
  <InvoiceLineInformation {...props} />,
  { wrapper: MemoryRouter },
));

describe('InvoiceLineInformation', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should display info invoice line connected to order line', () => {
    const { asFragment } = renderInvoiceLineInformation();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display info invoice line not connected to order line', () => {
    const { asFragment } = renderInvoiceLineInformation({
      ...defaultProps,
      poLine: {},
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
