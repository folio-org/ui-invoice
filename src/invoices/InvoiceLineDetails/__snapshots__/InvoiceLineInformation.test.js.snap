// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceLineInformation should display info invoice line connected to order line 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.description
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          EDI import invoice line
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.poLineNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <a
            class="root"
            data-test-text-link="true"
            href="/orders/lines/view/2e5067cd-2dc8-4d99-900d-b4518bb6407f"
            rel="noopener noreferrer"
            target="_blank"
          >
            1000-7
          </a>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.invoiceLineNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoice.details.information.status
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          Open
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionInfo
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionStart
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          2021-06-09
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionEnd
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          2021-06-18
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.comment
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.accountNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoice.accountingCode
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.quantity
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          20
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subTotal
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          $1,523.00
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="checkbox disabled"
        data-test-checkbox="true"
      >
        <label
          class="label label"
          for="checkbox-1"
        >
          <span
            class="inner disabledLabel"
          >
            <input
              aria-invalid="false"
              class="input"
              disabled=""
              id="checkbox-1"
              type="checkbox"
              value=""
            />
            <span
              class="checkboxIcon"
            >
              <svg
                height="10"
                viewBox="0 0 14 14"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.395 4.617L5.63 9.383 3.605 7.358"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-miterlimit="10"
                  stroke-width="1.5"
                />
              </svg>
            </span>
            <span
              class="labelText"
            >
              ui-invoice.invoiceLine.releaseEncumbrance
            </span>
          </span>
        </label>
        <div
          role="alert"
        />
      </div>
    </div>
    <div
      class="col-xs-12"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.referenceNumbers
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <div
            class="mclEmptyMessage"
            style="min-width: 200px;"
            tabindex="0"
          >
            stripes-components.tableEmpty
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InvoiceLineInformation should display info invoice line not connected to order line 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.description
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          EDI import invoice line
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.poLineNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span>
            -
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.invoiceLineNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoice.details.information.status
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          Open
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionInfo
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionStart
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          2021-06-09
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subscriptionEnd
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          2021-06-18
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.comment
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.accountNumber
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoice.accountingCode
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <span
            data-test-no-value="true"
          >
            <span
              class="sr-only"
            >
              stripes-components.noValue.noValueSet
            </span>
            <span
              aria-hidden="true"
            >
              -
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.quantity
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          20
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.subTotal
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          $1,523.00
        </div>
      </div>
    </div>
    <div
      class="col-xs-3"
    >
      <div
        class="checkbox disabled"
        data-test-checkbox="true"
      >
        <label
          class="label label"
          for="checkbox-2"
        >
          <span
            class="inner disabledLabel"
          >
            <input
              aria-invalid="false"
              class="input"
              disabled=""
              id="checkbox-2"
              type="checkbox"
              value=""
            />
            <span
              class="checkboxIcon"
            >
              <svg
                height="10"
                viewBox="0 0 14 14"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.395 4.617L5.63 9.383 3.605 7.358"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-miterlimit="10"
                  stroke-width="1.5"
                />
              </svg>
            </span>
            <span
              class="labelText"
            >
              ui-invoice.invoiceLine.releaseEncumbrance
            </span>
          </span>
        </label>
        <div
          role="alert"
        />
      </div>
    </div>
    <div
      class="col-xs-12"
    >
      <div
        class="kvRoot"
      >
        <div
          class="kvLabel"
        >
          ui-invoice.invoiceLine.referenceNumbers
        </div>
        <div
          class="kvValue"
          data-test-kv-value="true"
        >
          <div
            class="mclEmptyMessage"
            style="min-width: 200px;"
            tabindex="0"
          >
            stripes-components.tableEmpty
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
